sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2103,4 +2103,35 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       testData2.select(lit(7), 'a, 'b).orderBy(lit(1), lit(2), lit(3)),
       Seq(Row(7, 1, 1), Row(7, 1, 2), Row(7, 2, 1), Row(7, 2, 2), Row(7, 3, 1), Row(7, 3, 2)))
   }
+
+  test("SPARK-22226: splitExpressions should not generate codes beyond 64KB") {
+    val colNumber = 10000
+    val input = spark.range(2).rdd.map(_ => Row(1 to colNumber: _*))
+    val df = sqlContext.createDataFrame(input, StructType(
+      (1 to colNumber).map(colIndex => StructField(s"_$colIndex", IntegerType, false))))
+    val newCols = (1 to colNumber).flatMap { colIndex =>
+      Seq(expr(s"if(1000 < _$colIndex, 1000, _$colIndex)"),
+        expr(s"sqrt(_$colIndex)"))
+    }
+    df.select(newCols: _*).collect()
+  }
+
+  test("SPARK-22226: too many splitted expressions should not exceed constant pool limit") {
```

Comment:
I've just tried this. Seems we can simplify it to:
```scala
  test("SPARK-22226: too many splitted expressions should not exceed constant pool limit") {
    val colNumber = 6000
    val input = spark.range(2).rdd.map(_ => Row(1 to colNumber: _*))
    val df = sqlContext.createDataFrame(input, StructType(
      (1 to colNumber).map(colIndex => StructField(s"_$colIndex", IntegerType, false))))
    df.dropDuplicates((1 to 5).map(colIndex => s"_$colIndex")).collect()
  }
```
@mgaido91 Can you verify it? Thanks.

Reply:
this still fails for the same reason your first suggestion still fails

Reply:
Hmm, interesting. Because your committed test can be seen as to do `dropDuplicates` on a 6000 column dataframe (`df.select(funcs: _*)`). The `funcs` are actually performed at the lower `Project`.

```scala
    val df2 = df.select(funcs: _*).cache()
    df2.collect()  // Pass
    val df3 = df2.dropDuplicates((1 to 5).map(colIndex => s"_$colIndex"))
    df3.collect()  // Exception
```
So I think the simplified has the same effect?

Reply:
The following which can't be passed in current master branch can be passed with your fix. I didn't see OOM issue and nestedclass constant pool issue.

```scala
  test("SPARK-22226: too many splitted expressions should not exceed constant pool limit") {
    val colNumber = 5000
    val input = spark.range(2).rdd.map(_ => Row(1 to colNumber: _*))
    val df = sqlContext.createDataFrame(input, StructType(
      (1 to colNumber).map(colIndex => StructField(s"_$colIndex", IntegerType, false))))
    val funcs = (1 to colNumber).map { colIndex =>
      val colName = s"_$colIndex"
      col(colName).cast(LongType)
    }
    df.select(funcs: _*).dropDuplicates((1 to 5).map(colIndex => s"_$colIndex")).collect()
  }
```

Reply:
But actually I tested this committed test, once I lowered the threshold from 1600000 to 1000000, it can pass too in current master branch.

Reply:
It seems hard to make an end-to-end test for the reported issue after several tries.

Reply:
You are right @viirya. Sorry, I didn't notice. Yes the problem is that most of the times we have both these issues at the moment, thus solving one is not enough. It turns out that there are some corner cases in which this fix is enough, like the real case I am working on. But it is not easy to reproduce them in a simple way. In this use case there are a lot of complex projections a `dropDuplicate` and some joins after that. But there are query made of thousands of lines of SQL code.
The only way I have been able to reproduce it easily is in this test case: https://github.com/apache/spark/pull/19480/files#r144302922.

Reply:
The unit test added into `CodeGenerationSuite` looks sufficient for identifying this particular issue regarding constant pool limit in outer class due to too many method calls.

It is hard to contrive an end-to-end test so far purely for reproducing this particular issue. At least I failed to contrive one after several tries.

So let wait if anyone has the chance or insights to create one.

If no, I think the unit case in `CodeGenerationSuite` should be good enough.

Reply:
Btw, since this test didn't test what we want to test. We should remove it.